django/test/runner.py

```diff
@@ -468,6 +470,53 @@ def __iter__(self):
         return iter(self.subsuites)
 
 
+class Shuffler:
+    """
+    This class implements shuffling with a special consistency property.
+    Consistency means that, for a given seed and key function, if two sets of
+    items are shuffled, the resulting order will agree on the intersection of
+    the two sets. For example, if items are removed from an original set, the
+    shuffled order for the new set will be the shuffled order of the original
+    set restricted to the smaller set.
+    """
+
+    # This doesn't need to be cryptographically strong, so use what's fastest.
+    hash_algorithm = 'md5'
+
+    def __init__(self, seed):
+        self.seed = seed
+
+    @classmethod
+    def _hash_text(cls, text):
+        h = hashlib.new(cls.hash_algorithm)
+        h.update(text.encode('utf-8'))
+        return h.hexdigest()
+
+    def _hash_item(self, item, key):
+        text = '{}{}'.format(self.seed, key(item))
+        return self._hash_text(text)
```

Comment:
I don't see a reason we can't use Python's `hash()` builtin, which is even faster and cached on strings

I also don't think we need a class here - a single function to do the shuffling would do.

Reply:
> I don't see a reason we can't use Python's hash() builtin, which is even faster and cached on strings

There are a few related reasons. First, Python doesn't provide any guarantees about its `hash()` function. For example, it's implementation-dependent, and it can vary from Python version to Python version. Being different across implementations and versions could prevent one from being able to run the same test order across different versions. Also, since there aren't any guarantees, we don't know if the hash function used by the implementation actually mixes things well. Maybe it tends to keep certain strings in order, which isn't what we want. Using something originally designed for security like md5 (one of the "secure" hashes in Python's `hashlib` module) provides the uniform mixing we want.

Finally, one example of a change that CPython made to its hash() function is that it no longer gives the same values even from Python invocation to invocation. For example, see the note [here](https://docs.python.org/3/reference/datamodel.html#object.__hash__) in the Python docs for `object.__hash__()`:

> **Note:** By default, the `__hash__()` values of str and bytes objects are “salted” with an unpredictable random value. Although they remain constant within an individual Python process, they are not predictable between repeated invocations of Python.

You can see this from the command-line:

```
$ python
Python 3.9.4 (...) 
Type "help", "copyright", "credits" or "license" for more information.
>>> hash('abc')
5333155334811125317
>>> 
$ python
Python 3.9.4 (...) 
Type "help", "copyright", "credits" or "license" for more information.
>>> hash('abc')
-1048955330867350811
```

As for having a class, I think it helps to keep those methods grouped together, and it's nice to be able to pass an object around designed for the purpose, as opposed to e.g. some kind of function object. Is the issue more that you don't like passing an object around, or you don't like having the class at all? If the former, we could instead keep the class but pass around a callable with the right signature (e.g. a method on the object).